android/modules/ui/src/java/ti/modules/titanium/ui/widget/picker/TiUITimePicker.java

```diff
@@ -160,15 +174,36 @@ public void setValue(long value, boolean suppressEvent)
 	@Override
 	public void onTimeChanged(TimePicker view, int hourOfDay, int minute)
 	{
+		detectTimeChange();
+	}
+	
+	private void detectTimeChange() {
 		Calendar calendar = Calendar.getInstance();
-		calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
-		calendar.set(Calendar.MINUTE, minute);
-		if (!suppressChangeEvent) {
+		TimePicker picker = (TimePicker) getNativeView();
+		boolean hasChanged;
+		if (Build.VERSION.SDK_INT >= 23) {
+			hasChanged = lastSelectedHour != picker.getHour() || lastSelectMin != picker.getMinute();
```

Comment:
I wouldn't create public "ApiLevelX" classes because that shifts the responsibility to the user of our classes to implement the correct version fallback behavior. If we were to implement something like this, I think Google's "DrawableCompat" class (link below) would be a better way to do it, because it buries/hides the complexity from the users of our library/classes. With the one difference is that I wouldn't use dessert names like Google. I would name the classes based on API level number instead like I did above, which is less error prone. Especially since Google's API docs displays an API level number next to every API. (I've seen devs make bad assumptions about which dessert maps to which API level number, causing a crash.)
https://github.com/guardianproject/android-support-library/blob/master/v4/java/android/support/v4/graphics/drawable/DrawableCompat.java

But that said, my preference is to not create public bags of static functions under the AppCompat package/namespace. I think it makes more sense to group higher API levels calls by feature under the same package that uses them like "network", "ui", "view", etc. For example, a developer would normally expect all network related classes to be under the network namespace/package. Also, I prefer to keep the higher API level accessing classes as private inner-classes because this makes it more future proof. So, when we do raise the min API level we support in the future, we can refactor the obsolete API level accessing private inner-classes without it being a breaking change. This also means that I think wrapping Android features in our own classes is the way to go so that users of our classes don't have to worry about API levels, like how it works with many of Google's support library classes.

Regarding the TimePicker case above, since no other class in our code needs to call those 2 methods, it's simpler from a productivity standpoint to bury the higher level API accessor as an inner class like I've done it. This reduces the spaghetti code-ness of our library (it's only 14 lines of code). But if we ever need to rip it out in the future, then we can easily do so without it being a breaking change since it's a private inner class.

I don't expect us to fix all of these opcodes issues in one shot. These issues are everywhere and is too big of a job. Fixing these issues a bit at a time as we make changes is probably the way to go and we should strive to avoid adding new opcode warnings. I think this is a low priority issue since I've never heard of Google Play app reviewers reject apps due to opcode warnings (apps to be featured on Google Play actually go through an app review process). This is more of a tech-support issue on our end.

Oh and if you think the solution to solving these opcode warning is a hack, well, you'd be right. :-) But it's a Google sanctioned hack that Google's own support libraries 100% depend on.